rclcpp/src/rclcpp/node_options.cpp

```diff
@@ -338,7 +338,7 @@ NodeOptions::get_domain_id_from_env() const
   _dupenv_s(&ros_domain_id, &ros_domain_id_size, env_var);
 #endif
   if (ros_domain_id) {
-    uint32_t number = strtoul(ros_domain_id, NULL, 0);
+    uint32_t number = static_cast<uint32_t>(strtoul(ros_domain_id, NULL, 0));
```

Comment:
This static_cast just makes explicit what the compiler would also have done implicitly.

The check below may also not do what is expected: [`strtoul()`](https://en.cppreference.com/w/cpp/string/byte/strtoul) converts the string to `unsigned long`, and returns `ULONG_MAX` in case of an overflow. On my machine (Linux 64-bit) `unsigned long` is a 64 bit type, which is also why the implicit conversion to `uint32_t` triggered a compiler warning with `-Wconversion`. But that also means that only values exceeding `18446744073709551617` and all multiples of `4294967296` minus 1 trigger the run-time exception. `4294967296` is silently cast to `0`, `4294967297` to `1` etc.

Illegal values like `"foo"` would also silently be interpreted as `0` without an error, because `strtoul` simply stops parsing at the first invalid character. You would have to pass a `char **` as the second argument `str_end` instead of `NULL` and check whether it points to the end of the string (the terminating null character).

The conversions should probably use C++ [stringstreams](https://en.cppreference.com/w/cpp/io/basic_istringstream) or the C++11 equivalent [std::stoul](https://en.cppreference.com/w/cpp/string/basic_string/stoul) with an additional range check, instead of error-prone C functions. Or, if the function must be implemented in C to avoid code duplication with `rcl`, as stated in @wjwwood's comment above, the solution would be to use `unsigned long` instead of `uint32_t` without a cast, check that the `str_end` actually points to the terminating null character after conversion, and to add an explicit range check before the cast to the desired output type.

Apparently the issue is already discussed in https://github.com/ros2/rcl/pull/689.